File: auth/src/iosMain/kotlin/com/well/auth/credentialProviders/FacebookProvider.kt
package com.well.auth.credentialProviders

import cocoapods.FBSDKCoreKit.FBSDKApplicationDelegate
import cocoapods.FBSDKLoginKit.FBSDKLoginManager
import com.well.auth.Context
import com.well.auth.LoginView
import com.well.auth.credentialProviders.AuthCredential.FacebookCredential
import com.well.auth.resumeWithException
import kotlinx.coroutines.suspendCancellableCoroutine
import platform.Foundation.NSURL
import platform.UIKit.UIApplication
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

actual class FacebookProvider actual constructor(context: Context) : CredentialProvider(context) {
    private val loginManager = FBSDKLoginManager()

    init {
        context.apply {
            FBSDKApplicationDelegate.sharedInstance.application(
                application,
                launchOptions
            )
        }
    }

    actual override suspend fun getCredentials(loginView: LoginView): AuthCredential =
        suspendCancellableCoroutine { continuation ->
            loginManager.logInWithPermissions(
                listOf<String>(), loginView
            ) { result, error ->
                val token = result?.token?.tokenString()
                when {
                    error != null -> continuation.resumeWithException(error)
                    result?.isCancelled == true -> continuation.cancel()
                    token != null -> continuation.resume(FacebookCredential(token))
                    else -> continuation.resumeWithException(IllegalStateException())
                }
            }
        }

    override fun application(app: UIApplication, openURL: NSURL, options: Map<Any?, *>): Boolean =
        FBSDKApplicationDelegate.sharedInstance.application(
            app,
            openURL,
            options
        )
}

File: iosApp/Well/SharedPodImplementation/FacebookProvider.swift
//
//  FacebookProvider.swift
//  Well
//
//  Created by Philip Dukhov on 2/14/21.
//  Copyright © 2021 Well. All rights reserved.
//

import SharedMobile
import FBSDKCoreKit
import FBSDKLoginKit

final class FacebookProvider: CredentialProvider {
    private let loginManager: LoginManager

    override init(systemContext: SystemContext) {
        Settings.shared.isAdvertiserIDCollectionEnabled = true
        Settings.shared.loggingBehaviors = Set()
        ApplicationDelegate.shared.initializeSDK()
        Settings.shared.isAdvertiserIDCollectionEnabled = false
        loginManager = LoginManager()
        super.init(systemContext: systemContext)
    }

    override func getCredentials(completionHandler: @escaping (AuthCredential?, Error?) -> Void) {
        loginManager.logOut()
        loginManager.logIn(configuration: LoginConfiguration()!) { result in
            switch result {
            case let .success(_, _, token):
                if let token = token {
                    completionHandler(AuthCredential.FacebookCredential(token: token.tokenString), nil)
                } else {
                    completionHandler(nil, NSError(description: "Facebook login succeed but token is empty")/*KotlinIllegalStateException().toNSError()*/)
                }
            case .failed(let error):
                completionHandler(nil, error)

            case .cancelled:
                completionHandler(nil, KotlinCancellationException().toNSError())
            @unknown default:
                fatalError()
            }
        }
    }

    override func application(app: UIApplication, openURL: URL, options: [AnyHashable: Any] = [:]) -> Bool {
        ApplicationDelegate.shared.application(
            app,
            open: openURL,
            options: options as? [UIApplication.OpenURLOptionsKey: Any] ?? [:]
        )
    }
}
